### DISCUSSION TITLE
Segmentation fault when regridding

### DISCUSSION THREAD

**Discussion by AbhiHampiholi** [esmf-org/esmf#268] [Remapping / Regridding] [answered]
### Requirements

- [X] Reviewed [ESMF Reference Manual](https://earthsystemmodeling.org/doc/)
- [X] Searched [GitHub Discussions](https://github.com/orgs/esmf-org/discussions?discussions_q=)

### Affiliation(s)

AirMettle, Inc

### ESMF Version

8.6.0

### Issue

I am trying to use the ESMC library to regrid tensors obtained from reading NetCDF files. I am running into an issue where the first time I call regrid a file I get the correct expected output. But the second time, I get a segfault. I am running this with ESMF_COMM="mpiuni". 

I can provide more specifics but the general structure of the smallest failing example code is 

```
loop 2 times
    ESMC_Initialize(NULL, ESMC_InitArgLogKindFlag(ESMC_LOGKIND_NONE), ESMC_ArgLast);
    //create src grid - fill it up
    //create dest grid - fill it up
    //create src field using ESMC_FieldCreateGridTypeKind - fill up src field
    //create dest field
    //regrid between srcfield and destfield
    //read from destField to get "regridded data"
    ESMC_FieldDestroy(&srcfield);
    ESMC_FieldDestroy(&destfield);
    ESMC_GridDestroy(&srcgrid);
    ESMC_GridDestroy(&destgrid);
    ESMC_Finalize();
end loop
```

l have made sure to check that the return code is always = `ESMF_SUCCESS` after each call to an ESMF function. The first iteration of the loop runs perfectly and I get the correct expected output. However, on the second iteration, I get a segmentation fault in `ESMC_GridCreateNoPeriDim` The arguments that are being passed to `ESMC_GridCreateNoPeriDim` are identical on both iterations. On investigation with GDB I found that the cause of the segfault is that `petCount` is incorrectly set to a very large number (instead of 1) in the `ESMCI::DELayout::create()` function in the line
```
  // query the VM for localPet and petCount
  int petCount = vm->getNpets();
```
Any help finding the issue is much appreciated! 

### Autotag

@oehmke

**Comment by danrosen25** [comment, marked as answer]
@AbhiHampiholi 
We have this documented in the [ESMF Fortran Reference Guide](https://earthsystemmodeling.org/docs/release/latest/ESMF_refdoc/node4.html#SECTION04024500000000000000) but I see that it's not documented in the [ESMF C Reference Guide](https://earthsystemmodeling.org/docs/release/latest/ESMC_crefdoc/node4.html#SECTION04022200000000000000). I know that in Fortran you cannot call ESMF Initialize after Finalize. Try moving the ESMC_Initialize and ESMC_Finalize outside of the loop.

Reproducer
[test_268.tgz](https://github.com/user-attachments/files/16367929/test_268.tgz)

**Reply by AbhiHampiholi** [reply]
Thank you for the clarification @danrosen25 ! I confirmed that ensuring that no ESMC function gets called after the finalize fixes the issue.
